lsp/lsp.cpp

```diff
@@ -124,26 +136,48 @@ int main(int argc, const char **argv)
     // Parse that content as JSON
     JAST request;
     std::stringstream parseErrors;
-    if (!JAST::parse(content, parseErrors, request))
-    {
+    if (!JAST::parse(content, parseErrors, request)) {
       response.add("id", JSON_NULLVAL);
       JAST &error = response.add("error", JSON_OBJECT);
       error.add("code", JSON_INTEGER, ParseError);
       error.add("message", parseErrors.str());
     }
-    else
-    {
+    else {
       // What command?
       const std::string &method = request.get("method").value;
       const JAST &id = request.get("id");
-      //const JAST &params = request.get("params");
+      const JAST &params = request.get("params");
 
       // Echo back the request's id
       response.children.emplace_back("id", id);
 
-      JAST &error = response.add("error", JSON_OBJECT);
-      error.add("code", JSON_INTEGER, MethodNotFound);
-      error.add("message", "Method '" + method + "' is not implemented.");
+      if (isShutDown && (method != "exit")) {
+        JAST &error = response.add("error", JSON_OBJECT);
+        error.add("code", JSON_INTEGER, InvalidRequest);
+        error.add("message", "Received a request other than 'exit' after a shutdown request.");
+      }
+      else if (method == "initialize") {
+        response.children.emplace_back("result", initialize(params));
+      }
+      else if (!isInitialized) {
+        JAST &error = response.add("error", JSON_OBJECT);
+        error.add("code", JSON_INTEGER, ServerNotInitialized);
+        error.add("message", "Must request initialize first");
+      }
+      else if (method == "shutdown") {
+        JAST empty(JSON_OBJECT);
+        response.children.emplace_back("result", empty);
+        isShutDown = true;
+      }
+      else if (method == "exit") {
+        if (isShutDown) return 0;
```

Comment:
Since `return` is such an important directive, it should be highly visible, not hidden on the end of a line.
There are two ways you can achieve that:

```
if (isShutDown) {
  return 0;
} else {
  return 1;
}
```

Or, my preference: `return isShutdown?0:1;`
I'd say you should either p

Reply:
Done.